### PR TITLE
The New Auth Flow

### PR DESCRIPTION
**Pull request by jhuggett**
Implementing the new authentication flow which provides csrf mitigation and encrypts the Github api access token.

- [x] Update `react-tinacms-github` to work with the new flow

- [x] Modify the serverless function in `next-tinacms-github` to use the new flow

`react-tinacms-github` should require **no changes** to work. On its own, it handles both the new and old auth flow.

### Breaking Changes (`next-tinacms-github`):
`next-tinacms-github` api routes now requires a secret _Signing Key_.

The _Signing Key_ is just a random passphrase of your choice, a randomized 256-bit key is recommended. This key is only used server-side to encrypt and decrypt authentication tokens sent to the client.

One way to generate a recommended size key, is to run `openssl rand -base64 32` in your terminal and use the output as your _Signing Key_. 

`createAuthHandler`, `apiProxy`, and `previewHandler` now  **requires** the _Signing Key_ to be passed as a parameter. 

E.g., the proxy-github route:
```TSX
import { apiProxy } from 'next-tinacms-github'

export default apiProxy(
    process.env.SIGNING_KEY
)
```

**Also**, `enterEditMode` needs to pass the new token that is in local storage as an authorization header to the `/api/preview` route, like this:
```TSX
const enterEditMode = () => {
  const token = localStorage.getItem('token') || null

  const headers = new Headers()

  if (token) {
    headers.append('Authorization', 'Bearer ' + token)
  }


  return fetch(`/api/preview`, { headers: headers }).then(() => {
    window.location.href = window.location.pathname
  })
}
```

### CODE REVIEW

**Comment by ncphillips** [comment]
This looks good!

Could you update the PR description to list out the API breaking changes and describe how to migrate from the old flow to this one?

**Comment by ncphillips** [comment]
> requires the Signing Key to be passed as a parameter.

What is a "signing key" and how do I get one? Is just a random passphrase?

**Comment by ncphillips** [comment]
@dwalkr will that code generator require updates?

**Comment by jhuggett** [comment]
@ncphillips Yep, the signing key is just a random passphrase generated by the developer. I updated the description to explain it.

**Comment by dwalkr** [comment]
@nolan yes, just logged an issue https://github.com/dwalkr/hygen-next-tinacms-github/issues/3

File: packages/next-tinacms-github/src/github/proxy.ts [hunk 1]
```diff
@@ -15,24 +15,41 @@ See the License for the specific language governing permissions and
...
+        },
+      })
+        .then((resp: any) => {
+          res.status(resp.status).json(resp.data)
+        })
+        .catch((err: any) => {
+          res.status(err.response.status).json(err.response.data)
+        })
+    } else {
+      res.status(400).json('Bad Credentials')
```
Comment (dwalkr) [inline packages/next-tinacms-github/src/github/proxy.ts:50]: We should return better errors here

**Comment by ncphillips** [comment]
Awesome, we'll want to make sure to make that change and have it ready to merge when this gets released. 

@jhuggett the diff looks good I'm just going to run through our guide shortly

File: packages/react-tinacms-github/src/github-client/authenticate.ts [hunk 2]
```diff
@@ -34,17 +34,21 @@ export const authenticate = (
...
-          if (authTab) {
-            authTab.close()
-          }
-          resolve()
-        })
+        fetch(`${codeExchangeRoute}?code=${e.newValue}&state=${authState}`)
+          .then(response => response.json())
+          .then(data => {
+            const token = data.signedToken || null
+            if (token) {
```
Comment (dwalkr) [inline packages/react-tinacms-github/src/github-client/authenticate.ts:41]: Do you think we should plan to remove this check in the future? If so, we could add a deprecation warning in the console

**Comment by ncphillips** [comment]
Oh, does this also change it from an OAuth app to a GitHub app?

**Comment by dwalkr** [comment]
> Oh, does this also change it from an OAuth app to a GitHub app?

it does not

**Comment by ncphillips** [comment]
Running through the guide and I got stuck here.

I'm getting this error in the terminal:
```
API resolved without sending a response for /api/create-github-access-token?code=cf4e28df88131a7b9&state=vhf3r, this may result in stalled requests.
API resolved without sending a response for /api/proxy-github, this may result in stalled requests.
```

File: packages/react-tinacms-github/src/github-client/authenticate.ts [hunk 3]
```diff
@@ -41,6 +41,10 @@ export const authenticate = (
             if (token) {
               // for implementations using the csrf mitigation
               localStorage.setItem('token', token)
+            } else {
+              console.warn(
+                'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/the-new-auth-flow)'
```
Comment (ncphillips) [inline packages/react-tinacms-github/src/github-client/authenticate.ts:46]: Good to clarify it's just for GitHub
```suggestion
                'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/the-new-github-auth-flow)'
```

Comment (dwalkr) [inline packages/react-tinacms-github/src/github-client/authenticate.ts]:
```suggestion
                'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/upgrade-notice-tinacms-github-packages)'
```

File: packages/react-tinacms-github/src/github-client/github-client.ts [hunk 4]
```diff
@@ -333,6 +333,10 @@ export class GithubClient {
 
     if (token) {
       headers.append('Authorization', 'Bearer ' + token)
+    } else {
+      console.warn(
+        'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/the-new-auth-flow)'
```
Comment (dwalkr) [inline packages/react-tinacms-github/src/github-client/github-client.ts]:
```suggestion
        'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/upgrade-notice-tinacms-github-packages)'
```

File: packages/react-tinacms-github/src/github-client/github-client.ts [hunk 5]
```diff
@@ -326,8 +326,22 @@ export class GithubClient {
...
+    // For implementations using the csrf mitigation
+    const token = localStorage.getItem('token') || null
+
+    const headers = new Headers()
+
+    if (token) {
+      headers.append('Authorization', 'Bearer ' + token)
+    } else {
+      console.warn(
+        'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/the-new-github-auth-flow)'
```
Comment (dwalkr) [inline packages/react-tinacms-github/src/github-client/github-client.ts:338]:
```suggestion
        'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/upgrade-notice-tinacms-github-packages)'
```

File: packages/react-tinacms-github/src/github-client/authenticate.ts [hunk 6]
```diff
@@ -34,17 +34,25 @@ export const authenticate = (
...
+        fetch(`${codeExchangeRoute}?code=${e.newValue}&state=${authState}`)
+          .then(response => response.json())
+          .then(data => {
+            const token = data.signedToken || null
+            if (token) {
+              // for implementations using the csrf mitigation
+              localStorage.setItem('token', token)
+            } else {
+              console.warn(
+                'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/the-new-github-auth-flow)'
```
Comment (dwalkr) [inline packages/react-tinacms-github/src/github-client/authenticate.ts:46]:
```suggestion
                'Deprecation Notice: You are using an old authentication flow, please migrate to the new one (see https://tinacms.org/blog/upgrade-notice-tinacms-github-packages)'
```

**Pull request merged**